p4include/v1model.p4

```diff
@@ -133,14 +145,17 @@ parser Parser<H, M>(packet_in b,
                     inout standard_metadata_t standard_metadata);
 control VerifyChecksum<H, M>(in H hdr,
                              inout M meta);
+@pipeline
```

Comment:
I understand what you are trying to do.
The problem comes from the fact that you are trying to support multiple architectures with a single back-end. That will be harder and harder as these architecture diverge in behaviors. For example, changes you make for PSA may break v1model programs. Also, the BMv2 back-end generates a program called p4c-bm2-ss, so you see that it targets a specific simulator on top of BMv2 (bmv2 can be used to implements lots of different simulators), and it it assumes a specific model for the input architecture, v1model. These things are tightly coupled in the code; the C++ code hardwires a lot of knowledge about v1model and BMv2 SS in particular. You are trying to de-entangle them, but that won't be easy.

However, v1model should be thought as a description of a fixed architecture, which cannot be changed. In general, you cannot assume you can edit it anymore after you have distributed it to customers.

Moreover, regarding the @pipeline annotations, all the information you need IS in the package. It is just annoying to extract. Maybe you can write some helper functions to make this easier. For v1model we know that the pipelines are arguments 3 and 4 of the main package, and the deparser is argument 6 (the code in fact uses the parameter names in v1model to retrieve them). This is a contract between the architecture and the compiler, and it can never change once you published the architecture, so you can rely on it.  Similarly, for PSA you will identify the pipelines from the arguments to the main package.  So what you have to do is write a generic class that can identify the blocks from the main package, provide two implementations for PSA and v1model and output the same data-structure that you can use from both. I would suggest not using annotations, but just set<Node>, the same way it was done in the BMv2 back-end.